File: Data Structures & Algorithms/combination-target-sum-ii/submission-23.js
class Solution {
    /**
     * @param {number[]} candidates
     * @param {number} target
     * @return {number[][]}
     */
    combinationSum2(candidates, target) { 
        let result = []  

        candidates.sort()

        function dfs(j, cur, curSum) {
            if (curSum === target) {
                result.push([...cur]) 
                return 
            } 

            if (j >= candidates.length || curSum > target) {
                return
            } 
 

            for (let i = j; i < candidates.length; i ++) { 
                if (i > j && candidates[i] === candidates[i - 1]) continue
                cur.push(candidates[i]) 
                dfs(i + 1, cur, curSum + candidates[i]) 
                cur.pop()
            }
        } 

        dfs(0, [], 0) 

        return result
    }
}

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-23.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        numSum = sum(nums) 
        
        if numSum % 2 != 0: 
            return False 

        target = numSum // 2 

        def dfs(i, cur):    
            if i == len(nums): 
                return False 
                 
            if cur > target: 
                return False 

            if cur == target: 
                return True 

            take = dfs(i + 1, cur + nums[i]) 
            notTake = dfs(i + 1, cur) 

            return take or notTake 

        return dfs(0, 0)